### ISSUE TITLE
Sphinx Viewcode.py IndexError list index out of range  if child class file (.py) is smaller than parent class file.

### ISSUE THREAD

**Issue by hansonmcoombs** [sphinx-doc/sphinx#12026]
### Describe the bug

building sphinx the build fails if the child class is smaller than the parent class file.  

Error message

```
../sphinx/ext/viewcode.py", line 289, in collect_pages
    + lines[start])
      ~~~~~^^^^^^^
IndexError: list index out of range
```

Adding blank lines to the child class (at the end of the file) resolves the problem.... 

It seems like it must be a rather minimal problem, but it took some time to find the issue

### How to Reproduce

The easiest way to reproduce the error is to clone the demo repo I have created:  It has full environment information (conda) as well.  https://github.com/hansonmcoombs/sphinx_bug_test 

### Environment Information

```text
Platform:              linux; (Linux-5.15.0-94-generic-x86_64-with-glibc2.35)
Python version:        3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0])
Python implementation: CPython
Sphinx version:        7.2.6
Docutils version:      0.20.1
Jinja2 version:        3.1.3
Pygments version:      2.17.2
```

### Sphinx extensions

```python
['sphinx.ext.todo', 'sphinx.ext.viewcode', 'sphinx.ext.autodoc',
              'sphinx.ext.autosummary']
```

**Comment by jayaddison** [comment]
Hi @hansonmcoombs - thanks for taking the time to report this.  Can you check whether the problem occurs without the `autoapi` extension enabled?  It didn't for me, but it's possible I have a slightly different configuration somehow.

**Comment by hansonmcoombs** [comment]
@jayaddison  If I run it completely without the autoapi package then yes I get no error.  If I run with the auto api extension enabled I do get the error, but the rst files in the docs_build/autoapi directory are created.  If I then re-run with the auto api extension disabled (but with the auto api created rst files).  Everything compiles as expected.  Note, unsurprisingly there are some differences between the html output.  

Does this mean I should raise this as an error on https://github.com/readthedocs/sphinx-autoapi/issues?

Should I cross link it?

**Comment by jayaddison** [comment]
Thanks @hansonmcoombs.  Yep, if the issue is only reproducible with that extension enabled, then I think it's best to start there, and with a mention / hyperlink here so that the maintainer can catch up on the context so far.

I'd encourage trying to remove as much configuration and RST from your example as you can before doing that - your example is fairly minimal but it may be possible to trim some more and narrow it down (that might help you figure out what introduces the failure, and will certainly help the maintainer there track it down).

It may come back here if the bug turns out to be something about the internals of Sphinx or the API we provide -- but I think we should start at the extension that _seems_ to be the cause.

**Comment by hansonmcoombs** [comment]
@jayaddison Agreed and thanks for the advice.  I have found parsing the difference between "sphinx" and sphinx extensions a bit confusing, so a bit of clarity is much appreciated.

Re slimming down the repo; fair comment and have done.  I was at wits end when I posted the issue and that was as much as my brain could handle at the time. 

It appears to be a conflict
between autoapi and viewcode extensions, with the following settings:

- autoapi_options = ['inherited-members']
- autoapi_python_use_implicit_namespaces = True
- autoapi_python_class_content = 'both' 

I have pushed a commit with the slimmed down example.

For anyone else watching the new issue at autoapi is:  https://github.com/readthedocs/sphinx-autoapi/issues/422
